### PR TITLE
feat(sdk): add support for collecting system metrics from openmetrics endpoints

### PR DESCRIPTION
**Pull request by dmitryduev**
Description
-----------
Fixes WB-20680.

Adds support for collecting system metrics from OpenMetrics endpoints.

Example run: https://wandb.ai/dimaduev/open_metrics/runs/p1yicpc6

- [x] I updated CHANGELOG.md, or it's not applicable

Testing
-------
See the following report for more info: https://wandb.ai/dimaduev/dcgm/reports/Monitoring-GPU-cluster-performance-with-NVIDIA-DCGM-Exporter-and-Weights-Biases--Vmlldzo0MDYxMTA1

The following code snippet demonstrates how to configure OpenMetrics filters to capture only the metrics DCGM_FI_DEV_POWER_USAGE with the labels gpu="0" or gpu="1" from node1 and 2 metrics for all GPUs from node2.

```python
run = wandb.init(
    project="dcgm",
    settings=wandb.Settings(
        _stats_open_metrics_endpoints={
	    "node1": "http://192.168.0.1:9400/metrics",
	    "node2": "http://192.168.0.2:9400/metrics",
	},
	_stats_open_metrics_filters={
	    "node1.DCGM_FI_DEV_POWER_USAGE": {
	        "gpu": "[0,1]",
	    },
            "node2.DCGM_FI_DEV_(POWER_USAGE|GPU_TEMP)": {
	        "gpu": ".*",
	    },
	},
    ),
 )
```
The OpenMetrics endpoint at http://192.168.0.1:9400/metrics (designated `node1` by the user in the settings) might expose the following metrics:

```
# HELP DCGM_FI_DEV_POWER_USAGE Power draw (in W).
# TYPE DCGM_FI_DEV_POWER_USAGE gauge
DCGM_FI_DEV_POWER_USAGE{gpu="0",UUID="GPU-c601d117-58ff-cd30-ae20-529ab192ba51"} 99.9
DCGM_FI_DEV_POWER_USAGE{gpu="1",UUID="GPU-a7c8aa83-d112-b585-8456-5fc2f3e6d18e"} 0.1
DCGM_FI_DEV_POWER_USAGE{gpu="2",UUID="GPU-a7c8aa83-d112-b463-3456-5fc2f3f2k43s"} 40.2
```

The specified filter will capture only the metrics with lables gpu="0" and gpu="1" from node1.

### CODE REVIEW

File: core/pkg/monitor/openmetrics.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	"github.com/prometheus/common/expfmt"
+)
+
+// Filter represents a filter to apply to metrics collected from an OpenMetrics endpoint.
+//
+// Provides a fine-grained control over which metrics to collect.
+// OpenMetrics filters can be specified via wandb.Settings in one of the two formats:
+//   - {"metric regex pattern, including endpoint name as prefix": {"label": "label value regex pattern"}}
+//   - ("metric regex pattern 1", "metric regex pattern 2", ...)
+//
```
Comment (kptkin) [inline core/pkg/monitor/openmetrics.go:26]: it was hard for me to understand this, maybe a specific example would be helpful (or adding this documentation in the proto part- didn't check if it is there)

File: core/pkg/monitor/openmetrics.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+// to a list of Filter objects.
+func processMappingFilters(mapping *spb.MapStringKeyMapStringKeyStringValue) []Filter {
+	if mapping == nil {
+		return nil
+	}
+	result := []Filter{}
+	for metricRegex, labelFilters := range mapping.GetValue() {
+		filter := Filter{MetricNameRegex: metricRegex}
+		for labelName, labelRegex := range labelFilters.GetValue() {
+			filter.LabelFilters = append(filter.LabelFilters, [2]string{labelName, labelRegex})
```
Comment (kptkin) [inline core/pkg/monitor/openmetrics.go:56]: what do you think if it was a struct instead of an array with 2 values?

File: core/pkg/monitor/openmetrics.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+//
+// Supports only GAUGE and COUNTER metrics.
+type OpenMetrics struct {
+	name        string
+	url         string
+	filters     []Filter
+	client      *retryablehttp.Client
+	logger      *observability.CoreLogger
+	labelMap    map[string]map[string]int
+	labelHashes map[string]map[string]string
```
Comment (kptkin) [inline core/pkg/monitor/openmetrics.go:83]: some documentation of the different fields would be helpful, specifically:
- filters, labelMap, labelHashes (it's not obvious without reading the code what these are used for)

File: core/pkg/monitor/openmetrics.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+		labelHashes: make(map[string]map[string]string),
+	}
+
+	return om
+}
+
+func (o *OpenMetrics) Name() string { return o.name }
+
+// ShouldCaptureMetric checks if a metric should be captured based on the filters.
+func ShouldCaptureMetric(
```
Comment (kptkin) [inline core/pkg/monitor/openmetrics.go:137]: maybe having some caching mechanism would be good here, if i understand correctly this will match every time?

File: core/pkg/monitor/openmetrics.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+
+// Filter represents a filter to apply to metrics collected from an OpenMetrics endpoint.
+//
+// Provides a fine-grained control over which metrics to collect.
+// OpenMetrics filters can be specified via wandb.Settings in one of the two formats:
+//   - {"metric regex pattern, including endpoint name as prefix": {"label": "label value regex pattern"}}
+//   - ("metric regex pattern 1", "metric regex pattern 2", ...)
+//
+// These are converted to a list of Filter objects using the processMappingFilters
+// and processSequenceFilters functions, respectively.
```
Comment (kptkin) [inline core/pkg/monitor/openmetrics.go:29]: this seems implementation detail that could change, i would rephrase it

File: core/pkg/monitor/openmetrics.go [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+//	        "node1.DCGM_FI_DEV_POWER_USAGE": {
+//	            "gpu": "[0,1]",
+//	        },
+//	        "node2.DCGM_FI_DEV_(POWER_USAGE|MEM_COPY_UTIL|TOTAL_ENERGY_CONSUMPTION|GPU_TEMP|MEMORY_TEMP)": {
+//	            "gpu": ".*",
+//	        },
+//	    },
+//	),
+//
+// )
```
Comment (kptkin) [inline core/pkg/monitor/openmetrics.go:53]: maybe just describe the format of the input, i think this is too tied to what happens on the client side?

File: core/pkg/monitor/openmetrics.go [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+		return nil, fmt.Errorf("could not fetch metrics from endpoint")
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	var parser expfmt.TextParser
+	metricFamilies, err := parser.TextToMetricFamilies(resp.Body)
+	if err != nil {
+		o.logger.Error("Error parsing metrics", "error", err)
```
Comment (kptkin) [inline core/pkg/monitor/openmetrics.go:271]: could you add a prefix here for the function?

**Review by kptkin — APPROVED** [review]

File: core/pkg/monitor/openmetrics.go (hunk 7 above)
Comment (dmitryduev) [reply core/pkg/monitor/openmetrics.go:271]: can you explain please?

Comment (kptkin) [reply core/pkg/monitor/openmetrics.go:271]: Oh just the regular format we use in other messages;
function name: error-message

Comment (dmitryduev) [reply core/pkg/monitor/openmetrics.go]: ah, got it, sry
